### DISCUSSION TITLE
Register Read/Write through Home Assistant

### DISCUSSION THREAD

**Discussion by jupppo** [johanmeijer/grott#235] [Show and tell]
Hi all,

I am about to built a homeassistant page to read out the registers from datalogger and inverter, getting a meaningful description step by step and finally update the register data through HA Webinterface.
First Stage is ready and looks like this:
![Clipboard01](https://user-images.githubusercontent.com/69680107/201983940-0e35515e-cea3-40e0-b40c-f83e211550c4.jpg)
One minor problem is that I get a response time out from grott server, when I pull too much data at the same time (working on this).
The pic show the first 20 registers of the datalogger.

For those who want to play by themself. This is the entry in configuration.yaml:

```#SENSORS
sensor:

# grottstatus via GET
  - platform: rest
    name: growatt_dl_register_1
    json_attributes:
      - value
    resource: http://192.168.1.140:5782/datalogger?command=register&register=1&datalogger=XXXXX
    method: GET
    headers:
      Content-Type: application/json
    payload: '{"register": "value"}'
    value_template: "{{ value_json.success }}"
    scan_interval: 120
  - platform: template
    sensors:
      gwreg1:
        friendly_name: "Growatt DL Register 1"
        value_template: "{{ state_attr('sensor.growatt_dl_register_1', 'value')}}" ```

**Comment by Chreutz** [comment]
Maybe the problem is the low baud rate of the inverter. AFAIK, it only operates at 9600, so you can very quickly flood the queue. A way to limit the request rate might be the answer.

**Reply by rany2** [reply]
Not true, my fork of Grott could read/write registers in under a second (often in <600ms) but it depends on whether a data record is being sent or not. If a data record was just sent, it could take a second or two. Link is https://github.com/rany2/grott

**Reply by jupppo** [reply]
I agree. I timed the requests differntly, but I still get some registers and some getting a time out first 6 are always there, but the rest is shuffeled. Very strange, CPU and Memory is fine on the server. Same for the inverter registers.

**Comment by johanmeijer** [comment]
Reading the inverters register is a little bit instabel (understatement). That is true voor using the grottserer and also growatt website. I am not sure where that is coming from. It sometimes takes seconds (that is why I wait a long time in the grottserver) and sometimes there is a wrong answer (0) or no answer at all (timeout). Make it faster helps only if the inverter/datalogger replies faster.   

We can shorten the wait time, that is not a problem. It only raises the chance on no answer (getting a timeout). 
I will make it at least configurable in a next release of the grottserver. 

I can not repair or mask the instable inverter reply on the inverter register read / write requests in grottserver I think (maybe it is possible with some kind of request scheduling / cache mechanism but that would be a huge change with limit improvement).

By the way I think that write is more stable dan the read.

**Reply by rany2** [reply]
Yes, for me the write has always worked without fail when the inverter returns a confirmation. On the latest data logger firmware, the return value is extremely large so we might just want to consider it a failure if the register return value is greater than 20,000 for example
